Fixes #4629 - Migrate UI to Patternfly

alongoldboim opened the pull request.

ohadlevy commented:
when viewing facts (monitor -> facts) sorting the table based on value shows empty screen.

ohadlevy commented:
i take it back, it was an error on my browser-sync setup.

ohadlevy commented:
icon colors are blue, not sure if its on purpose, but imho black background looks better.
also, if we switch to font-awesome, their background would be black.

ohadlevy commented:
under host edit, all inherit buttons are not aligned

alongoldboim commented:
@ohadlevy @dLobatog fixed the issues above, have a look and ill be glad to get some more comments.

ohadlevy commented:
@alongoldboim parameters add btn should be btn-primary vs btn-success.

ohadlevy commented:
when clicking on template preview, the select2 size does not match

ohadlevy commented:
tooltip colors are not always consitent, e.g in the hostgroup edit page (see attached)

![tooltip-color](https://cloud.githubusercontent.com/assets/20167/11627503/9055ddba-9cf3-11e5-927b-226fa734fa21.png)

ohadlevy commented:
add matcher (under smart / class params) still uses btn-success maybe grep for all btn-success and figure out if they should be switched to primary.

ohadlevy commented:
host modal error still line wrap incorrectly.

![host-new-interfaces](https://cloud.githubusercontent.com/assets/20167/11628218/f539f946-9cf8-11e5-8101-06520ba7efd4.png)

ohadlevy commented:
@alongoldboim do you mind updating what did you changed? its hard to know which comments were addressed.

alongoldboim commented:
@ohadlevy sure thing, first of all i fixed the failing tests (some js concerning global parameters was counting on having a button with btn-success and we changed it to primary so i modified the script), made some other fixes you didn't address, next commit will deal with all the comments above, and ill update from now on each commit.

alongoldboim commented:
@ohadlevy Regarding the btn-success comments, in the current UI all those buttons current class is btn-success, would you like to change this to btn-primary app wide for all buttons with adding functionality?

alongoldboim commented:
As example over here, changing the add variable button doesn't looks so good in my opinion.
Maybe i should change the tab above ("new") it to be white colored?
![selection_006](https://cloud.githubusercontent.com/assets/11405684/11678467/ebd210ba-9e50-11e5-9210-d2d4930e24a5.png)

dLobatog commented on app/assets/stylesheets/patternfly_and_overrides.scss:
```diff
@@ -3,72 +3,78 @@
-.nav li.dropdown-submenu .dropdown-menu:after{
-  display: none;
+body {
+  font-size: 14px;
```
The default recommended font-size by Patternfly is 12px, @ohadlevy @alongoldboim , [see their test pages](https://rawgit.com/patternfly/patternfly/master/tests/basic.html), the idea is to use patterns that have been user-tested, provide a common UI framework for enterprise apps and not deviate much from it as we're not going to do any testing. A simple example that comes to mind, [foreman_cockpit components](https://camo.githubusercontent.com/dd327f958993d60ecefbe4e1fc26bc45ad63cf9c/687474703a2f2f692e696d6775722e636f6d2f527a64735239622e676966) would not fit as well if we don't follow Patternfly recommendations.

dLobatog commented on app/assets/stylesheets/application.scss:
```diff
@@ -236,6 +235,7 @@ table {
 
   tr td:last-child {
     .btn {
+      font-size: 13px;
```
This makes the edit buttons in forms stick out a bit too much. It sticks out as much as a column title, and more than a primary button text. Also the sub elements on a dropdown menu look weird.

![screenshot from 2015-12-09 10-02-43](https://cloud.githubusercontent.com/assets/598891/11680941/0bfd2cf2-9e5c-11e5-8032-5cb4c0f84f1b.png)

If we respect the default body font 12px for Patternfly and remove this font-size it looks more consistent in my opinion. 

![screenshot from 2015-12-09 10-27-25](https://cloud.githubusercontent.com/assets/598891/11681486/81862674-9e5f-11e5-94e3-4a247cc3d3ef.png)

@kybaker any comments on how the Edit button should look like?

dLobatog replied:
Another problem with this is that most other text sizes have to be changed to match. Notice how the title is smaller than the regular font if we override it to 14px and by using the default it's just fine. 

![screenshot from 2015-12-09 10-53-26](https://cloud.githubusercontent.com/assets/598891/11682174/8b7a048a-9e63-11e5-8ae5-f86f71a12ccb.png)
![screenshot from 2015-12-09 10-53-53](https://cloud.githubusercontent.com/assets/598891/11682175/8da4f080-9e63-11e5-8dba-3013e02f3cec.png)

dLobatog commented on app/assets/stylesheets/patternfly_and_overrides.scss:
```diff
@@ -3,72 +3,78 @@
-.dropdown-submenu {
-  position:relative;
+.form-control{
+  font-size: 14px;
```
Ditto as above

alongoldboim commented:
@dLobatog @ohadlevy I think the major thing over here is to decide if we want to follow patternfly 12px or we want to use our regular 14px css, it effects everything.

dLobatog commented:
@alongoldboim Thanks for taking this. I've closed the original PR. I found these problems, aside from the general font problem that affects everything. One of these problems can be fixed by using the default Patternfly font of 12px, maybe others can be fixed that easily too. Let me know what you think.

1) Images under Host#new operating system are not aligned with the rest of the form 
![screenshot from 2015-12-09 10-34-07](https://cloud.githubusercontent.com/assets/598891/11681651/82c6e37e-9e60-11e5-9f0f-c9962c2d15f6.png)

2) Font to add config groups is not standard. 
![screenshot from 2015-12-09 10-39-34](https://cloud.githubusercontent.com/assets/598891/11681752/2ddfb20e-9e61-11e5-9d09-f256c8eb4cae.png)

3) Spinner for Domain and Subnet shows up in the middle of the text, it should show up next to the field. Notice 'Error generating IP: Not found' stack up and show up below the IP address field. They should show up in the field.
![screenshot from 2015-12-09 10-41-34](https://cloud.githubusercontent.com/assets/598891/11681843/a5c8ded0-9e61-11e5-9229-71c70931a996.png)
![screenshot from 2015-12-09 10-43-03](https://cloud.githubusercontent.com/assets/598891/11681846/a82eb988-9e61-11e5-9208-b4d5306c9c31.png)

4) Parameters table headers are wider than they should, try to follow https://rawgit.com/patternfly/patternfly/master/tests/basic.html and just use h2 for `Puppet class parameters` and `Global parameters` instead of legend (that's just a leftover from Bootstrap 2 I think). Notice the border-radius of the button to override puppet classes should be 0px (pointy, not round).

Same problem on the matchers table of any Puppet class smart variables, Domain parameters, etc..

![screenshot from 2015-12-09 10-47-34](https://cloud.githubusercontent.com/assets/598891/11681938/45ae6276-9e62-11e5-8fdd-490e26ca9b53.png)

5) 'Add all' button on selectable container is glued to the search filter, and the list is narrower than the title. This one fixes itself by changing body to the default of 12px.
![screenshot from 2015-12-09 10-58-16](https://cloud.githubusercontent.com/assets/598891/11682266/f52f4624-9e63-11e5-99ad-b631806f5bd0.png)
![screenshot from 2015-12-09 11-11-04](https://cloud.githubusercontent.com/assets/598891/11682493/9b2f27a0-9e65-11e5-82a1-4229c3088d96.png)

6) Merge/avoid duplicates check boxes are tilted to the right on Puppet class smart variables. Also the whole form looks very weird due to the body font change to 14px

![screenshot from 2015-12-09 11-04-09](https://cloud.githubusercontent.com/assets/598891/11682369/9893cfd8-9e64-11e5-8c15-c751a5098a95.png)

@kybaker should subtitles in Forms use h2/h3 rather than legend here? 

7) Compute profiles loading spinner when you edit a compute profile on a compute resource is not the font-awesome spinner, it's still the gif that should be removed. LDAP test connection also uses the old spinner.gif.

8) Long fact values don't wrap properly, see this example
![screenshot from 2015-12-09 11-08-34](https://cloud.githubusercontent.com/assets/598891/11682442/41ac8826-9e65-11e5-925d-d23166eee61b.png)

alongoldboim commented:
ignore latest commit, adding commit using PF original font size and making the needed fixes around that.

ohadlevy commented:
can you please rebase.

dLobatog commented:
@alongoldboim Sorry, I don't see any fixes for any of above comments 1 - 8? https://github.com/theforeman/foreman/pull/2943#issuecomment-163176003

alongoldboim commented:
@dLobatog Had to deal with some hammer business, all of the above will be done by Monday :).

alongoldboim commented:
@ohadlevy @dLobatog Made a few changes:
- changed compute profile edit page label to use col-md-2 instead of 3 to fit the rest of the ui forms.
- changed tool tips to be black colored.
- fixed templates preview select2 size.
- changed btn-success to btn-primary for adding/new functionality.
- host edit interfaces modal errors are shown under the field correctly. 
- changed default font to 12 px (lets trust pf :) ).
- changed edit buttons tables to use default font size.

Regarding Daniel's comments:
1) fixed.
2) problem with changing the font because it ruins the plus sign, can be fixed but tbh even when  changing the font is almost identical.
3) spinner fixed, as for the error messages i decided to place them under the fields because some of them are long and overflow, check it and let me know if you are OK with it.
4) changed to h2, we need to decide if that under-line is important, it can be added.
5) fixed.
6) fixed, check boxes are closer to the labels.
7) fixed spinner for all two-pane modals.
8) as for what I have seen the facts are truncated appropriately.

dLobatog commented:
@alongoldboim Thanks, good fixes :smile:  . A few more minor things:

1) Select2 boxes text is not centered. I tried changing line-height on patternfly_overrides for select2-container select2-choice to 26px or 2.2 and it seems centered, but I haven't looked if it breaks anything else.
![screenshot from 2015-12-15 10-48-43](https://cloud.githubusercontent.com/assets/598891/11807331/71c4f526-a319-11e5-96a7-34a91c07038f.png)
![screenshot from 2015-12-15 10-48-50](https://cloud.githubusercontent.com/assets/598891/11807346/744943e2-a319-11e5-8291-de4e5eec689d.png)
2) The IP address and errors are fine I guess, the problem is that they show up multiple times. If you select a wrong subnet several times, it will create several "Error generating IP errors" without clearing the previous ones. If you click on suggest new, it does clear the previous errors.
3) Parameters table heads should be white, not this light gray color.
4) Parameters actions column is not centered 
![screenshot from 2015-12-15 11-03-26](https://cloud.githubusercontent.com/assets/598891/11807674/8b9933ac-a31b-11e5-88c0-665118ed38b1.png)
5) Fact values still don't wrap, notice the difference 
![screenshot from 2015-12-15 11-29-53](https://cloud.githubusercontent.com/assets/598891/11808206/487966a6-a31f-11e5-8c12-9e9261352567.png)
![](https://cloud.githubusercontent.com/assets/598891/11682442/41ac8826-9e65-11e5-925d-d23166eee61b.png)
6) Merge overrides labels are still shifted to the right. The idea is to make the labels aligned with the order label. We have the same kind of problem above that with "Use puppet default" and "Hidden value" See this:
![screenshot from 2015-12-15 11-32-48](https://cloud.githubusercontent.com/assets/598891/11808619/cb2ff1ee-a321-11e5-84f1-9d5a61255d4c.png)

dLobatog commented:
Aside from that, could you try not vendoring assets like Patternfly and Font-awesome and using https://rubygems.org/gems/patternfly-sass instead? It wasn't there when I started the original PR, but it seems like we can do this without putting everything under /vendor :smile:

alongoldboim commented:
@dLobatog regarding your last comment:
1) done.
2) done.
3+4) didn't quite get those comments, all tables headers are using this light gray color (would you like to change them all?), all actions columns buttons aren't centered (same question).
5) done.
6) done.

About using patternfly-sass gem, ill do this change after moving to rails4.

dLobatog commented:
@alongoldboim Thanks for the update. Rails 4 is in and you need to rebase it, but don't worry, from what I saw there are only 2 minor conflicts.

About 3) - the problem is that without Patternfly the table head is white, so that in Parameters it doesn't look weird and it fits. With Patternfly i'd argue the parameters table head looks very strange & without borders in light gray, all other tables are fine.

4) The actions column of parameters is centered (class 'ca'), so it looks misaligned - @tbrisker fixed this a while ago IIRC. I think the actions column should be centered as well in other tables, but we can leave that for another PR. 
![screenshot from 2015-12-18 15-29-28](https://cloud.githubusercontent.com/assets/598891/11898985/439cd4cc-a59c-11e5-9930-dd95fe23c2ff.png)
![screenshot from 2015-12-18 15-41-59](https://cloud.githubusercontent.com/assets/598891/11899232/e4723332-a59d-11e5-8b71-12ffa1622775.png)

Other than these 2 changes & using 'patternfly-sass' instead of vendoring the assets, I think everything looks fine. Try to be on the lookout for any regressions though :smile: 

Thanks @alongoldboim !

alongoldboim commented:
@dLobatog @ohadlevy Removed vendored fonts, moved to use pattenfly-sass gem.

alongoldboim commented:
Fixed table header in parameter tab to be white (comment 3).

dLobatog commented on bundler.d/assets.rb:
```diff
@@ -1,7 +1,8 @@
 gem 'jquery-ui-rails', '< 5.0.0'
-gem 'bootstrap-sass', '3.0.3.0'
+gem 'bootstrap-sass'
+gem 'patternfly-sass'
```
`patternfly-sass` relies on `bootstrap-sass`, feel free to remove the latter from here. Also please pin it to a stable version, `~> 2.7` should be fine

dLobatog replied:
We need to pin gem 'autoprefixer-rails', '~> 5.2' here, 6.0+ is only Ruby 2.0 compatible and we still support 1.9.3

dLobatog commented on LICENSE:
```diff
@@ -27,6 +27,7 @@ The following files and directories are exceptions:
 * lib/tasks/convert.rake is under Revised BSD license
 * extras/noVNC/websockify is LGPL v3 licensed
 * vendor/assets/fonts/ is ASL 2.0 licensed
+* vendor/assets/fonts/font-awesome is SIL OFL 1.1 licensed
```
No need to include this, patternfly.css, nor the LICENSE text if we don't vendor them :smile:

dLobatog commented:
@alongoldboim I think the reason both rubocop & all 1.9.3 tests failed is that bundler tries to install the latest autoprefixer-rails, fix it with this - https://github.com/theforeman/foreman/pull/2943#discussion_r48228993

alongoldboim committed:
Fixes #4629 - Migrate UI to Patternfly

alongoldboim commented:
@dLobatog @ohadlevy LGTM, check it please.

ohadlevy commented:
I'm seeing the following

```
2015-12-22T22:42:08 [app] [W] DEPRECATION WARNING on line 30 of foreman/app/assets/stylesheets/gridster.scss: #{} interpolation near operators will be simplified
 | in a future version of Sass. To preserve the current behavior, use quotes:
 | 
 |   unquote("#{percentage($i / 12)} - #{$col-margin-fallback}")
 | 
 | You can use the sass-convert command to automatically fix most cases.
 | 
```

ohadlevy commented:
I'm also seeing 404 in the logs for missing fonts? (or maybe they were renamed?)
e.g.
 GET http://0.0.0.0:3000/fonts/bootstrap/glyphicons-halflings-regular.ttf 404 (Not Found)

ohadlevy commented:
actually, i think it my local branch fault, ignore (404 on fonts)

alongoldboim commented:
@ohadlevy will check this dep call.

dLobatog commented on LICENSE:
```diff
@@ -1106,3 +1107,101 @@ Library.
     of your accepting any such warranty or additional liability.
 
 -----------------------------------------------------------------------------
+
```
No need for this license if it comes through a gem, we only include licenses for stuff we directly vendor in the repo

dLobatog replied:
@alongoldboim ^

dLobatog replied:
unfixed

dLobatog commented:
@alongoldboim Almost there, there are a few issues:

1) If you enable a plugin that has assets, for instance foreman-docker, it throws the following error 
 | Sprockets::FileNotFound - couldn't find file 'bootstrap' with type 'application/javascript':
2) The header of matchers table color doesn't match the form, same problem as with parameters before. 

![screenshot from 2015-12-23 13-52-12](https://cloud.githubusercontent.com/assets/598891/11976839/798e4032-a97c-11e5-9d8a-2b390faf6bea.png)

3) The actions column is severely misaligned with the 'remove' matcher button.
![screenshot from 2015-12-23 13-53-23](https://cloud.githubusercontent.com/assets/598891/11976855/95b059f8-a97c-11e5-97eb-cbb5b397d3c5.png)

Apparently notifications on Patternfly 2.x changed a bit, https://www.patternfly.org/patterns/toast-notifications/ and https://www.patternfly.org/patterns/inline-notifications/, we should think about fixing that at some point.

I'll start working on the rpm spec. Thanks!

ehelms commented:
I began testing the removal of Patternfly from Bastion to use it from this pull request as preparation. Using Katello, I hit the following error:

```
 | ActionView::Template::Error (couldn't find file 'bootstrap' with type 'application/javascript'):
 |     16:     <%= yield(:stylesheets) %>
 |     17: 
 |     18:     <%= csrf_meta_tags %>
 |     19:     <%= javascript_include_tag 'application', "data-turbolinks-track" => true %>
 |     20:     <%= yield(:javascripts) %>
 |     21: 
 |     22:     <script type="text/javascript">
 |   app/assets/javascripts/application.js:8
 |   app/views/layouts/base.html.erb:19:in `_403de3fd76eaf57e72e2a5f235245f17'
 |   app/views/layouts/login.html.erb:10:in `_c645e44f4e8245c4e0898946529a9661'
 |   app/controllers/users_controller.rb:85:in `login'
 |   app/controllers/concerns/application_shared.rb:13:in `set_timezone'
 |   app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
 |   lib/middleware/catch_json_parse_errors.rb:9:in `call'
```

To begin isolating the issue, I disabled Katello and enabled only foreman-docker and hit the same issue. Same issue with foreman-tasks. In fact, this error is true of any plugin that follows our standard template. I  dug through code, commented code out and arrived at the culprit as being our use of `:after => :finisher_hook` when registering the plugin with Foreman (e.g. https://github.com/theforeman/foreman-docker/blob/master/lib/foreman_docker/engine.rb#L42). After I changed this to a `config.after_initialize` the error went away.

What was the actual error? From what I could discern, certain asset gems were being removed or never added to the `config.assets.paths` list and therefore Sprockets never knew about them. For example, if you look at the `config.assets.paths` when Sprockets loads (https://github.com/rails/sprockets-rails/blob/2.x/lib/sprockets/railtie.rb#L100) with foreman-docker only, you get the following list:

```
["/home/vagrant/foreman/app/assets/images",
 "/home/vagrant/foreman/app/assets/javascripts",
 "/home/vagrant/foreman/app/assets/stylesheets",
 "/home/vagrant/foreman/vendor/assets/fonts",
 "/home/vagrant/foreman/vendor/assets/javascripts",
 "/home/vagrant/foreman/vendor/assets/stylesheets",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/jquery_pwstrength_bootstrap_4-1.2.2/vendor/assets/javascript",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/spice-html5-rails-0.1.5/vendor/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/flot-rails-0.0.3/vendor/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/ace-rails-ap-4.0.1/vendor/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/gettext_i18n_rails_js-1.0.3/lib/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/gettext_i18n_rails_js-1.0.3/vendor/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/underscore-rails-1.8.3/vendor/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/select2-rails-3.5.9.3/vendor/assets/images",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/select2-rails-3.5.9.3/vendor/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/select2-rails-3.5.9.3/vendor/assets/stylesheets",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/gridster-rails-0.5.6/vendor/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/gridster-rails-0.5.6/vendor/assets/stylesheets",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/multi-select-rails-0.9.12/app/assets/images",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/multi-select-rails-0.9.12/app/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/multi-select-rails-0.9.12/app/assets/stylesheets",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/jquery-ui-rails-4.2.1/app/assets/images",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/jquery-ui-rails-4.2.1/app/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/jquery-ui-rails-4.2.1/app/assets/stylesheets",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/jquery-turbolinks-2.1.0/vendor/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/jquery-rails-3.1.4/vendor/assets/javascripts",
 "/home/vagrant/foreman-docker/app/assets/javascripts",
 "/home/vagrant/foreman-docker/app/assets/stylesheets",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/turbolinks-2.5.3/lib/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/scoped_search-3.2.2/app/assets/images",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/scoped_search-3.2.2/app/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/scoped_search-3.2.2/app/assets/stylesheets"]
```

Now if we switch the finisher_hook to `config.after_initialize`, the output looks much better:

```
"/home/vagrant/foreman/app/assets/images",
 "/home/vagrant/foreman/app/assets/javascripts",
 "/home/vagrant/foreman/app/assets/stylesheets",
 "/home/vagrant/foreman/vendor/assets/fonts",
 "/home/vagrant/foreman/vendor/assets/javascripts",
 "/home/vagrant/foreman/vendor/assets/stylesheets",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/jquery_pwstrength_bootstrap_4-1.2.2/vendor/assets/javascript",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/spice-html5-rails-0.1.5/vendor/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/flot-rails-0.0.3/vendor/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/ace-rails-ap-4.0.1/vendor/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/gettext_i18n_rails_js-1.0.3/lib/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/gettext_i18n_rails_js-1.0.3/vendor/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/underscore-rails-1.8.3/vendor/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/select2-rails-3.5.9.3/vendor/assets/images",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/select2-rails-3.5.9.3/vendor/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/select2-rails-3.5.9.3/vendor/assets/stylesheets",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/gridster-rails-0.5.6/vendor/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/gridster-rails-0.5.6/vendor/assets/stylesheets",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/multi-select-rails-0.9.12/app/assets/images",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/multi-select-rails-0.9.12/app/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/multi-select-rails-0.9.12/app/assets/stylesheets",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/jquery-ui-rails-4.2.1/app/assets/images",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/jquery-ui-rails-4.2.1/app/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/jquery-ui-rails-4.2.1/app/assets/stylesheets",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/jquery-turbolinks-2.1.0/vendor/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/jquery-rails-3.1.4/vendor/assets/javascripts",
 "/home/vagrant/foreman-docker/app/assets/javascripts",
 "/home/vagrant/foreman-docker/app/assets/stylesheets",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/turbolinks-2.5.3/lib/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/scoped_search-3.2.2/app/assets/images",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/scoped_search-3.2.2/app/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/scoped_search-3.2.2/app/assets/stylesheets",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/bootstrap-sass-3.3.6/assets/stylesheets",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/bootstrap-sass-3.3.6/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/bootstrap-sass-3.3.6/assets/fonts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/bootstrap-sass-3.3.6/assets/images",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/font-awesome-sass-4.3.2.1/assets/stylesheets",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/font-awesome-sass-4.3.2.1/assets/fonts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/patternfly-sass-2.7.0/assets/stylesheets",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/patternfly-sass-2.7.0/assets/javascripts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/patternfly-sass-2.7.0/assets/fonts",
 "/home/vagrant/.rvm/gems/ruby-2.2.4/gems/patternfly-sass-2.7.0/assets/images"]
```

I haven't determined WHY the finisher_hook method leads to this issue, only that changing it to after_initialize fixes the issue and seems like the most appropriate hook based on (http://guides.rubyonrails.org/v4.1.5/configuring.html#using-initializer-files) where in the stack plugins were loaded previously and the available config hooks.

I feel this change (or whatever is decided as the appropriate change) should go into the major plugins prior to merging this else they will all be broken. I am more than happy to help with this effort. I wanted to bring it to light for discussion before proceeding.

dLobatog commented:
Just for context - https://github.com/Katello/katello/pull/5687/files & https://github.com/Katello/bastion/pull/88

alongoldboim commented:
@dLobatog @ehelms You think that changing to `config.after_initialize` is the right solution?
If so we need to make PR for all of the plugins, plus for the plugin-template.

alongoldboim commented:
Pinned down sprockets version.

dLobatog commented:
@alongoldboim I'd rather not break more APIs than what Rails 4 did already, I'm investigating what's exactly going on and why sprockets is not able to load the engine assets.

ohadlevy commented on app/assets/javascripts/application.js:
```diff
@@ -234,6 +240,7 @@ function template_info(div, url) {
     },
     error: function(jqXHR, textStatus, errorThrown) {
       $(div).html('<div class="alert alert-warning alert-dismissable">' +
+        '<span class="pficon-layered"><span class="pficon pficon-warning-triangle"></span><span class="pficon pficon-warning-exclamation"></span></span>' +
```
does it make sense to have an icon helper? similar to https://github.com/theforeman/foreman/pull/2971/files#diff-a9c3bd311eab80c9ebe6a69830f9ad02R560

alongoldboim replied:
We can even use the same one, but just need to add kind to it so we will be able to use pficon and fa, not only glyph. should i add this function in this patch also?
return (span class="#{kind}-' + name + '" )

ohadlevy replied:
i think the proxy one will be merged first as its nearly ready,then you could rebase on top of it.

alongoldboim replied:
already created such a function and applied it to all needed sections.

ohadlevy commented on app/assets/javascripts/application.js:
```diff
@@ -390,24 +397,40 @@ function foreman_url(path) {
+  switch(type)
+  {
+  case 'success':
+    return "<span class='pficon pficon-ok'></span><strong>" + __('Success') + ": </strong>";
```
icon function dry :)

ohadlevy commented on app/assets/javascripts/editor.js:
```diff
@@ -127,6 +127,37 @@ function create_editor() {
   }
 }
 
+function set_fullscreen(){
```
where does this comes from? is it required for pf ?

alongoldboim replied:
@dLobatog this comes from your pr? any particular reason or mistake?

dLobatog replied:
@alongoldboim IIRC this was needed before we started to use the Ace editor, it doesn't look like it's used anywhere now

alongoldboim replied:
@dLobatog should i remove it?

tbrisker replied:
The real `set_fullscreen` that is actually used is in `application.js`. This looks like remains from before the editor was disconnected from the fullscreen button. Same goes for the `exit_fullscreen`function below.

dLobatog replied:
@alongoldboim I think it can be removed yeah

ohadlevy commented on app/assets/javascripts/host_edit.js:
```diff
@@ -543,7 +543,8 @@ function image_provision_method_selected() {
 $(document).on('change', '#host_provision_method_image', image_provision_method_selected);
 
 $(document).on('change', '.interface_domain', function () {
-  interface_domain_selected(this);
```
nitpick: indentation is off

ohadlevy commented on app/assets/stylesheets/application.scss:
```diff
@@ -219,13 +220,20 @@ select {
+    padding-left: 10px;
+  }
+}
+
```
nitpick: whitespacing and extra line

ohadlevy commented on app/assets/stylesheets/application.scss:
```diff
@@ -282,23 +290,23 @@ table {
 
   .sorting {
-    background: image-url('sort_both.png') no-repeat center right;
+    background: url('sort_both.png') no-repeat center right;
```
should we replace this with an icon maybe?

ohadlevy replied:
actually, i don't think we even use it?

alongoldboim replied:
Checked it and I don't think we are.
removing.

tbrisker replied:
this is actually used by data tables (for example, in compute resource->show->virtual machines tab).

alongoldboim replied:
kk, didn't see that, added them back..

tbrisker replied:
maybe to make sure this isn't missed as easily and reduce the size of app.scss, extract these classes to data-tables.scss

ohadlevy replied:
:+1:

ohadlevy commented on app/assets/stylesheets/application.scss:
```diff
@@ -340,7 +348,7 @@ table {
   width: 16px;
   height: 16px;
-  background: image-url('spinner.gif');
+  background: url('spinner.gif');
```
spinner.gif was deleted?

shlomizadok replied:
In Rails4 no `image-url` afaik

ohadlevy replied:
no, i mean the file is actually deleted in this pr. (and you are right about the url vs image-url).

alongoldboim replied:
This whole css class isn't used anymore.
In the next commit it will be removed.

ohadlevy replied:
it is used in two-pane and also in #2971. we would have to rebase one or the other depends on which one gets merged first.

alongoldboim replied:
We would need to rebase indeed, spinner_placeholder function was modified, this css class is no longer relevant.

ohadlevy commented on app/assets/stylesheets/bootstrap_and_overrides.scss:
```diff
@@ -1,248 +0,0 @@
-}
-
-// Necessary because bootstrap3 nav dropdown does not have top arrow
-.nav li .dropdown-menu:after {
```
dont we need this anymore?

ohadlevy commented on app/assets/stylesheets/login.scss:
```diff
@@ -65,7 +58,7 @@
 }
 .login-page .container .form-horizontal .control-label {
-  font-size: 12px;
+  font-size: 13px;
```
didn't we agree on 12px?

ohadlevy commented on app/assets/stylesheets/patternfly_and_overrides.scss:
```diff
@@ -0,0 +1,347 @@
+  white-space: normal;
+  max-width: 500px;
+}
+
```
whitespace

ohadlevy commented on app/assets/stylesheets/patternfly_and_overrides.scss:
```diff
@@ -0,0 +1,347 @@
+  padding-right: 5px;
+}
+
+div.spinner{
```
some time spinner is in a span?

ohadlevy commented on app/assets/stylesheets/patternfly_and_overrides.scss:
```diff
@@ -0,0 +1,347 @@
+}
+
+.form-control{
+  display: inline-block  //font-family: 'Open Sans', Helvetica, Arial, sans-serif;
```
remove comment?

ohadlevy commented on app/assets/stylesheets/patternfly_and_overrides.scss:
```diff
@@ -0,0 +1,347 @@
+  padding-bottom: 0px;
+}
+
+td i{
```
what does 'i' mean?

alongoldboim replied:
@ohadlevy @tbrisker In Layout_helper.rb we got icon_text function that we use to define our icons with and it returns a <i> tag.
We can switch 'i' with 'span' (http://stackoverflow.com/questions/11135261/should-i-use-i-tag-for-icons-instead-of-span).

tbrisker replied:
+1 to replacing all `i` icons with `span`. `i` originally meant italic, and in [HTML5](http://www.w3.org/TR/html5/text-level-semantics.html#the-i-element) means this text has a different tone. The usage for icons is an abuse of the element simply because it is short, starts with the same letter and is not used very often.

alongoldboim replied:
though Bootstrap toke the <i> tag and using it officially as their icon tag.
@ohadlevy If we decide on replacing all icons to use span instead of the <i> tag we should do it on a diff pr, should i open an issue?

tbrisker replied:
Bootstrap 3 doesn't use `<i>`, it looks like this is a left over from bootstrap 2. 
This can be a different issue, but any changes here should keep that in mind (i.e use `span` instead of `i`).
In any case I would remove any reference to the `i` element from the css, and use only class matching. If wee need both `glyphicon` and `pficon` classes, we can add another class to all that is `icon` for css that is common to both.

ohadlevy replied:
:+1:

ohadlevy replied:
actually, font awesome uses `<i>` AFAIR?  e.g.:
http://fortawesome.github.io/Font-Awesome/icon/minus-circle/

tbrisker replied:
Interesting, they indeed use `<i>` in their examples, but they [say that it isn't semantically correct](http://fortawesome.github.io/Font-Awesome/examples/#basic). 
I checked their source code everything is done by class only, nowhere do they refer to the `i` element.

ohadlevy commented on app/assets/stylesheets/patternfly_and_overrides.scss:
```diff
@@ -0,0 +1,347 @@
+}
+
+.autocomplete-clear{
+  font-size: 14px
```
14px or 12px?

ohadlevy commented on app/assets/stylesheets/patternfly_and_overrides.scss:
```diff
@@ -0,0 +1,347 @@
+  background-color: black;
+}
+
+//.input-group-btn > button{
```
commented out on purpose?

ohadlevy commented on app/assets/stylesheets/patternfly_and_overrides.scss:
```diff
@@ -0,0 +1,347 @@
+//}
+
+span.input-group-btn .hidden-xs {display:inline !important;}
+@media (max-width: 768px) {
```
nitpick: odd char before media?

tbrisker replied:
that's the way css media selectors work, this is correct.

ohadlevy commented on app/helpers/layout_helper.rb:
```diff
@@ -329,20 +330,6 @@ def form_to_submit_id(f)
     "aid_#{key}_#{model}"
   end
 
-  def submit_or_cancel(f, overwrite = false, args = { })
```
why is this moved from the file?

alongoldboim replied:
The file was too long and rubocop generated a too long error so i moved it to application.

ohadlevy replied:
:-1: this is where it should be.

ohadlevy replied:
unresolved.

ohadlevy commented on app/helpers/layout_helper.rb:
```diff
@@ -425,14 +412,31 @@ def alert(opts = {})
+  def alert_header(text, html_class = nil)
+    case html_class
+      when /alert-success/
+        icon = "<span class='pficon pficon-ok'></span>"
```
icon helper?

ohadlevy commented on app/views/auth_source_ldaps/index.html.erb:
```diff
@@ -4,24 +4,24 @@
 <% title_actions(display_link_if_authorized(_("New LDAP Source"), hash_for_new_auth_source_ldap_path), documentation_button('4.1.1LDAPAuthentication')) %>
 
-<table class="table table-bordered table-striped table-two-pane">
+<table class="table table-bordered table-striped table-two-pane table-hover">
```
use `table_css_classes` helper?

ohadlevy commented on app/views/bookmarks/index.html.erb:
```diff
@@ -1,19 +1,19 @@
 <div class="title_action"></div>
 
-<table class="table table-bordered table-striped table-two-pane">
+<table class="table table-bordered table-striped table-two-pane table-hover">
```
ditto

ohadlevy commented on app/views/common_parameters/index.html.erb:
```diff
@@ -3,22 +3,23 @@
-        <td class="display-two-pane ellipsis"><%= link_to_if_authorized common_parameter, hash_for_edit_common_parameter_path(:id => common_parameter).merge(:auth_object => common_parameter, :authorizer => authorizer, :permission => 'edit_globals') %></td>
-        <td class="ellipsis"><%= common_parameter.safe_value %></td>
+        <td class="display-two-pane ellipsis"><%= link_to_if_authorized trunc_with_tooltip(common_parameter), hash_for_edit_common_parameter_path(:id => common_parameter).merge(:auth_object => common_parameter, :authorizer => authorizer, :permission => 'edit_globals') %></td>
+        <td class='ellipsis' style='overflow-wrap: break-word;'><%= trunc_with_tooltip(common_parameter.safe_value, 80) %></td>
```
why hardcode style here vs a css class?

ohadlevy replied:
i also thought we stopped truncing with ruby, @tbrisker ?

tbrisker replied:
`trunc_with_tooltip` should not be used in fixed width table cells with `ellipsis` class. I am not sure why the `table-fixed` class was removed from this table?

ohadlevy commented on app/views/hosts/review_before_build.html.erb:
```diff
@@ -1,11 +1,9 @@
                      :id                          => 'recheck_review', :data => { :url => review_before_build_host_path(:id => @host) }) %>
-    </h4>
-
+                   </br>
```
why `<br>`? would wrapping in `<p>` would be better?

ohadlevy commented on app/views/images/index.html.erb:
```diff
@@ -8,7 +8,7 @@
       <th><%= s_("Image|Uuid") %></th>
       <th><%= s_("Image|User data") %></th>
-      <th></th>
+      <th><<%= _('Actions') %>/th>
```
extra "<"

ohadlevy commented on app/views/trends/index.html.erb:
```diff
@@ -2,22 +2,9 @@
-    <%= alert_close %>
-    <p><strong><%= _("No trend counter defined.") %></strong></p>
-    <p><%= _("To define trend counters, use the Add Trend Counter button.") %><br/>
-      <%= (_("To start collecting trend data, set a cron job to execute 'foreman-rake trends:counter' every Puppet Interval (%s minutes).") % Setting.puppet_interval).html_safe %></p>
```
why remove the first warning?

alongoldboim replied:
@dLobatog this comes from your pr? any particular reason or mistake?

dLobatog replied:
I don't think anything was removed here? The if condition now comprises both warnings, that's it.

ohadlevy replied:
@dLobatog the whole text is missing... note these are different messages based on different conditions.
1. create new trends (you dont have any)
2. collect trend data (collection is not running).

ohadlevy replied:
unresolved.

alongoldboim replied:
To start collecting trend data, set a cron job to execute 'foreman-rake trends:counter' every Puppet Interval (%s minutes).
To start collecting trend data, set a cron job to execute <span class='black'> foreman-rake trends:counter </span> every Puppet Interval (%s minutes).

@ohadlevy failing to see the diff in messages, the only diff is in the header, but because the solution is running the same command I'm not sure that it would make a difference, i can revert it to be the same as before, let me know.

ohadlevy replied:
On Sun, Jan 3, 2016 at 10:33 AM, alon goldboim notifications@github.com
wrote:

> > - <p><%= _("To define trend counters, use the Add Trend Counter
> >   button.") %><br/>

^^ this line.

ohadlevy commented on vendor/assets/javascripts/jquery.jnotify.js:
```diff
@@ -160,7 +160,8 @@
 		this.remove = function (callback){
-			var $msg = $note.find("." + options.classMessage), $parent = $msg.parent();
+
+			var $msg = $note.find("." + options.classMessage.replace(" ",",")), $parent = $msg.parent();
```
I'm not happy with changing vendored javascript, two questions:
1. dont we have pf alterts js built in?
2. do we need to update jnotify?

why do we need this change at all?

alongoldboim replied:
@dLobatog this comes from your pr, whats the reason for the change.

dLobatog replied:
Without this change you can't close dismissable alerts. Moving to pure bootstrap/patternfly alerts would fix it too, but it might be a big change. Maybe you can do that on a separate PR (before this), as you wish. I don't like changing the vendored js either.

alongoldboim replied:
Lets get this pr in first, then we can change the alerts and remove this vendor change.

ohadlevy replied:
I would prefer before :)

ohadlevy replied:
if we change where we call it, it might not be required to change it here.

ohadlevy commented:
@alongoldboim please also note test failures. thanks

ohadlevy commented on app/views/hosts/_selected_hosts.html.erb:
```diff
@@ -1,5 +1,5 @@
 <div class="row">
-  <div class="col-md-8">
+  <div class="col-md-12">
```
if its 12, its the entire grid? do we need the extra div then?

alongoldboim commented:
@ohadlevy @dLobatog 
- @dLobatog was added in comments regarding his changes.
- changed notify function to work with new pf icons.
- added a new icon helper function for js, and applied it.
- removed spinner-placeholder scss class, not used anymore.
- fixed minor issues, indentations,spacing.
- sorting classes, didn't move to icons yet.
- bootstrap_and_overrides was deleted and not needed anymore, the required overrides were moved to patternfly_and_overrides.
- removed irrelevant font-size from scss.
- removed font-size on autocomplete-clear making it default to 16px, found that looking better.
- found a lot of places to use the table_css_classes helper, opened a separated issue.(http://projects.theforeman.org/issues/12945)
- switched icon helper to use span instead of i tag.
- removed unnecessary br tag wrapped with p tag instead .
- removed unnecessary div col12.
- moved submit_or_cancel function back to layout_helper.
- header of matchers table color changed to white in order to fit the form better.
- matchers actions column aligned.

ohadlevy commented on app/assets/stylesheets/application.scss:
```diff
@@ -333,19 +341,11 @@ table {
 }
 
-.spinner-placeholder {
-  width: 16px;
```
is there an alternative to setting up a spinner via javascript?

alongoldboim replied:
The spinner-placeholder function still exists just modified it a bit, we got the PF spinner class built in, so this css class is no longer needed.

alongoldboim commented:
Failing due to "module has too many lines" on layout helper, we can either change rubocop's settings for higher max-lines, or move stuff?

dLobatog commented on .byebug_history:
```diff
@@ -0,0 +1,11 @@
+c
```
I guess this was committed by mistake

dLobatog commented on bundler.d/assets.rb:
```diff
@@ -10,8 +13,8 @@
 gem 'flot-rails', '0.0.3'
-gem 'sass-rails', '~> 4.0'
 gem 'uglifier', '>= 1.0.3'
+gem 'sass-rails', '~> 5.0.4'
```
`~> 5.0` ought to be fine, this is pinning it to `5.0.9` max, minor versions shouldn't break compatibility. Same above for sprockets-rails

dLobatog commented:
@alongoldboim Thanks for the updates. The pr needs to be rebased, I left a few comments inline explaining a couple of changes you or @ohadlevy didn't understand. I don't get any see rubocop errors on LayoutHelper though?

ohadlevy commented on bundler.d/assets.rb:
```diff
@@ -1,7 +1,10 @@
 gem 'jquery-turbolinks', '~> 2.1'
 gem 'jquery-ui-rails', '< 5.0.0'
-gem 'bootstrap-sass', '3.0.3.0'
+gem 'autoprefixer-rails', '~> 5.2'
```
imho this should be a seperate PR to discuss assets compilation in general. :-1:

dLobatog replied:
That one is needed for the rest of the PR to work, it's a dependency of bootstrap-sass and if it's not pinned it won't work on 1.9.3

ohadlevy commented on bundler.d/assets.rb:
```diff
@@ -1,7 +1,10 @@
 gem 'jquery-ui-rails', '< 5.0.0'
-gem 'bootstrap-sass', '3.0.3.0'
+gem 'autoprefixer-rails', '~> 5.2'
+gem 'sprockets', '~> 2.12'
```
AFAIU  you need at least 2.8 to work with sass-rails https://rubygems.org/gems/sass-rails

ohadlevy commented on bundler.d/assets.rb:
```diff
@@ -1,7 +1,10 @@
+gem 'autoprefixer-rails', '~> 5.2'
+gem 'sprockets', '~> 2.12'
+gem 'patternfly-sass', '~> 2.7'
+gem 'sprockets-rails', '~> 2.3.3'
```
requires sprockets >3.0 https://rubygems.org/gems/sprockets-rails

tbrisker commented on app/assets/stylesheets/patternfly_and_overrides.scss:
```diff
@@ -3,75 +3,105 @@
 $navbar-default-bg: $primary_color;
 
-@import "bootstrap";
+.fact-col{
```
Why is this here? This file should only contain overrides to the default patternfly css. 
Any css that is not related to overriding the defaults should either go:
a. in case it applies to every/most pages - in application.scss
b. in case it applies to a specific page - in <page>.scss, scoped under an id selector inside that page if relevant.

tbrisker commented on app/assets/stylesheets/patternfly_and_overrides.scss:
```diff
@@ -3,75 +3,105 @@
+  float: left;
+}
+
+.spinner-two-pane{
```
-> two_pane.scss

tbrisker commented on app/assets/stylesheets/patternfly_and_overrides.scss:
```diff
@@ -3,75 +3,105 @@
-.navbar .nav > li > .dropdown-menu.pull-right:after {
-  left: auto;
-  right: 13px;
+.select2-arrow > b{
```
-> application.scss

tbrisker commented on app/assets/stylesheets/patternfly_and_overrides.scss:
```diff
@@ -3,75 +3,105 @@
-  margin-left:-1px;
-  max-height: 420px;
-  overflow-y: auto;
+.ms-container .ms-selectable li.ms-elem-selectable, .ms-container .ms-selection li.ms-elem-selection{
```
-> application.scss

tbrisker commented on app/assets/stylesheets/patternfly_and_overrides.scss:
```diff
@@ -3,75 +3,105 @@
-  border-left-color:#cccccc;
-  margin-top:5px;
-  margin-right:-10px;
+.ms-filter{
```
you get the point...

tbrisker commented on app/assets/stylesheets/puppetclasses_or_envs_changed.scss:
```diff
@@ -0,0 +1,6 @@
+.popover-content {
```
I'm pretty sure this was removed already in the past...

tbrisker commented on app/assets/javascripts/application.js:
```diff
@@ -58,10 +58,16 @@ function onContentLoad(){
+    if (!$(item).hasClass('btn-info') && !$(item).hasClass('btn-danger')) {
+      $(item).addClass("btn-default");
+    }
+  });
```
The last 5 lines can be written much more clearly and easily as
`$("#title_action a").addClass('btn').not('.btn-info, .btn-danger').addClass("btn-default");`

tbrisker commented on app/assets/javascripts/application.js:
```diff
@@ -556,3 +578,9 @@ function disableButtonToggle(item, explicit) {
 }
+
+function icon_text(icon_class, name, inner_text) {
+  var icon = '<span class="' + icon_class + "-" + name + '"/>'
```
Why not add the icon_class twice instead of passing "pficon pficon" all the time? 
Also please try to make the js function and the rb function that do this more similar, the order is reversed between them making it harder to remember which parameter is which.

tbrisker commented on app/assets/javascripts/jquery.multi-select.js:
```diff
@@ -5,8 +5,8 @@ $(function(){
-      selectableHeader: $("<div class='ms-header'>" + __('All items') + " <input placeholder='" + __('Filter') + "' class='ms-filter' type='text'><a href='#' title='" + __('Select All') + "' class='ms-select-all pull-right glyphicon glyphicon-plus icon-white'></a></div>"),
-      selectionHeader: $("<div class='ms-header'>" + __('Selected items') + "<a href='#' title='" + __('Deselect All') + "' class='ms-deselect-all pull-right glyphicon glyphicon-minus icon-white'></a></div>"),
+      selectableHeader: $("<div class='ms-header'>" + __('All items') + " <input placeholder='" + __('Filter') + "' class='ms-filter' type='text'><a href='#' title='" + __('Select All') + "' class='ms-select-all pull-right glyphicon glyphicon-plus'></a></div>"),
+      selectionHeader: $("<div class='ms-header'>" + __('Selected items') + "<a href='#' title='" + __('Deselect All') + "' class='ms-deselect-all pull-right glyphicon glyphicon-minus'></a></div>"),
```
According to patternfly guidelines, glyphicons are available but should be replaced by pficons or font-awesome when possible. (for example, here we can use fa-plus/fa-minus)

ohadlevy replied:
:+1:

ohadlevy replied:
http://projects.theforeman.org/issues/12972

alongoldboim commented:
@dLobatog @ohadlevy 
- rebased according to the proxy pr.
- Moved some of the css to the relevant css file.
- changed to use sass import instead of sprockets require (works well)
- removed popover-content.
- change sprockets gem version to be >3 .

ohadlevy commented:
DEPRECATION WARNING: Extra .css in SCSS file is unnecessary. Rename foreman/vendor/assets/stylesheets/vendor.css.scss toforeman/vendor/assets/stylesheets/vendor.scss. 

the same for vendor/assets/stylesheets/jquery.jnotify.scss.

ohadlevy commented:
DEPRECATION WARNING on line 23 of foreman/app/assets/stylesheets/gridster.scss: #{} interpolation near operators will be simplified

alongoldboim commented:
@ohadlevy
- removed extra css from jquery.jnotify and vendor.
- talked with @tbrisker and didn't find a solution for it, any thoughts?

tbrisker commented on app/assets/javascripts/application.js:
```diff
@@ -558,7 +575,8 @@ function disableButtonToggle(item, explicit) {
-function icon(name) {
-  "use strict";
-  return '<span class="glyphicon glyphicon-' + name + '" />';
+function icon_text(name, icon_class, inner_text) {
```
This still doesn't match the ruby version that uses (name, text, class) for the parameter order.

ohadlevy replied:
also, why drop use_strict?

tbrisker commented on app/helpers/layout_helper.rb:
```diff
@@ -413,7 +400,8 @@ def icons(i)
 
   def icon_text(i, text = "", opts = {})
-    (content_tag(:i,"", :class=>"glyphicon glyphicon-#{i} #{opts[:class]}", :title => opts[:title]) + " " + text).html_safe
+    opts[:kind] = "glyphicon glyphicon" if opts[:kind].nil?
```
I think we should default to pf, not glyphicon?

ohadlevy replied:
can be written as

``` ruby
opts[:kind] ||= "glyphicon glyphicon"
```

ohadlevy replied:
also, if its truly type / kind, it should be only glyphicon of pficon, no point in passing the string twice.

alongoldboim replied:
will change it.

tbrisker commented on app/helpers/layout_helper.rb:
```diff
@@ -413,7 +400,8 @@ def icons(i)
   def icon_text(i, text = "", opts = {})
-    (content_tag(:i,"", :class=>"glyphicon glyphicon-#{i} #{opts[:class]}", :title => opts[:title]) + " " + text).html_safe
+    opts[:kind] = "glyphicon glyphicon" if opts[:kind].nil?
+    (content_tag(:span,"", :class=>"#{opts[:kind]}-#{i} #{opts[:class]}", :title => opts[:title]) + " " + text).html_safe
```
duplicate `#{opts[:kind]}` here instead of passing the kind twice

ohadlevy replied:
text doesnt work with pficon, its not the same layout.

alongoldboim replied:
@ohadlevy could you explain, what do you mean by text doesn't work?

ohadlevy replied:
Try using the same method with pf font and text.

tbrisker commented on app/views/common_parameters/index.html.erb:
```diff
@@ -3,22 +3,23 @@
 
-<table class="table table-bordered table-striped table-two-pane table-fixed">
+
+<table class="table table-bordered table-striped table-two-pane table-hover">
```
Why remove the table-fixed class here and use trunc_with_tooltip below?

alongoldboim replied:
@dLobatog you made this change, any specific reason ?

dLobatog replied:
I think when I did it trunc_with_tooltip was the only option because the table wouldn't do it, so feel free to revert it

ohadlevy commented:
any ideas for better UX for the reports page? the labels are not as easy to parse imho now
![pf-reports](https://cloud.githubusercontent.com/assets/20167/12051333/dd6f4072-af0a-11e5-8d41-1c584f425adc.png)

ohadlevy commented on .gitignore:
```diff
@@ -35,4 +35,5 @@ tags
 _build
 zeus.json
 custom_plan.rb
+.byebug_history
```
not related?

ohadlevy commented on app/assets/javascripts/application.js:
```diff
@@ -392,24 +393,40 @@ function foreman_url(path) {
 function spinner_placeholder(text){
   if (text == undefined) text = "";
-  return "<div class='spinner-placeholder'>" + text + "</div>"
+  return "<div class='spinner-two-pane'> <div id='Loading' class='spinner spinner-md spinner-inline'> </div>" + text + "</div>";
```
what does two-pane has to do with spinner placeholder?

ohadlevy commented on app/assets/javascripts/proxy_status.js:
```diff
@@ -2,11 +2,11 @@ function setItemStatus(item, response) {
     item.attr('title', __('Active'));
     item.addClass('text-success');
-    item.html(icon('ok-sign'));
+    item.html(icon_text('ok-sign', "glyphicon"));
```
same comment as above?

ohadlevy commented on app/assets/stylesheets/patternfly_and_overrides.scss:
```diff
@@ -239,23 +223,61 @@ textarea {
   }
 }
 
+.glyphicon-info-sign{
```
doesn't this change for all usage of the info sign?

alongoldboim replied:
Yes, do you see a problem with it?

ohadlevy replied:
not sure, can you verify (search for all places with info sign..) maybe the classes selections etc?

ohadlevy commented on app/assets/stylesheets/two_pane.scss:
```diff
@@ -1,13 +1,14 @@
+  top: 50%;
+  position: absolute;
+  left: 45%;
+}
```
note you have spinner placeholder in this same file a few lines bellow..

ohadlevy commented on app/helpers/application_helper.rb:
```diff
@@ -307,10 +307,11 @@ def select_action_button(title, options = {}, *args)
-          content_tag(:ul,:class=>"dropdown-menu pull-right") do
-            args.map{|option| content_tag(:li,option)}.join(" ").html_safe
-          end
+      link_to((title + " " + content_tag(:i, '', :class => 'caret')).html_safe, '#',
```
did we agree to keep `<i>` vs `<span>` ?

ohadlevy commented on app/helpers/layout_helper.rb:
```diff
@@ -502,4 +513,27 @@ def spinner
+  def spinner(text = '', options = {})
+    if text.present?
+      "<div id='#{options[:id]}' class='spinner spinner-xs spinner-inline #{options[:class]}'>
+      </div> #{text}".html_safe
```
imho the text should be on the left side and the spinner on the right side?

ohadlevy commented on app/helpers/layout_helper.rb:
```diff
@@ -502,4 +513,27 @@ def spinner
+    end
+  end
+
+  def submit_or_cancel(f, overwrite = false, args = { })
```
remove from here.

ohadlevy commented on app/views/bookmarks/welcome.html.erb:
```diff
@@ -1,8 +1,12 @@
+  <p><%= _("You don't seem to have any bookmarks.") %></p>
+  <p><%= _("Bookmarks are saved search queries. To save them, click on 'Bookmark this search' on the dropdown next to every search button.") %></p>
+  <div class="blank-slate-pf-main-action">
+      <%= link_to _('Documentation'), "http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/index.html#4.1.5Searching", :rel => "external", :class => "btn btn-primary btn-lg" %>
```
should use documentation_url helper

alongoldboim commented on app/assets/stylesheets/application.scss:
```diff
@@ -1,14 +1,14 @@
+@import "jquery.ui.core";
+@import "jquery.ui.menu";
+@import "jquery.ui.theme";
+@import "jquery.ui.autocomplete";
```
@ohadlevy @dLobatog 
the 3 imports above are required by autocomplete, when importing //=require function won't work, so i just imported all of the needed files myself.

ohadlevy commented on app/views/common/500.html.erb:
```diff
@@ -1,15 +1,16 @@
-<% title _("Oops, we're sorry but something went wrong") %>
-
-
-<%= alert :class => 'alert-block alert-danger base in fade', :text => h(exception.message) %>
```
should we add an icon? maybe like http://fontawesome.io/icon/frown-o/

ohadlevy commented on app/views/common_parameters/index.html.erb:
```diff
@@ -3,22 +3,23 @@
       <tr>
-        <td class="display-two-pane ellipsis"><%= link_to_if_authorized common_parameter, hash_for_edit_common_parameter_path(:id => common_parameter).merge(:auth_object => common_parameter, :authorizer => authorizer, :permission => 'edit_globals') %></td>
-        <td class="ellipsis"><%= common_parameter.safe_value %></td>
+        <td class="display-two-pane ellipsis"><%= link_to_if_authorized trunc_with_tooltip(common_parameter), hash_for_edit_common_parameter_path(:id => common_parameter).merge(:auth_object => common_parameter, :authorizer => authorizer, :permission => 'edit_globals') %></td>
```
we mentioned this before, trunc with tooltip should not be used.

ohadlevy commented on app/views/compute_resources/form/_rackspace.html.erb:
```diff
@@ -5,4 +5,4 @@
                  :help_inline => link_to_function(_("Test Connection"), "testConnection(this)",
                  :class => "btn + #{flavors.empty? ? "btn-default" : "btn-success"}",
-                 :'data-url' => test_connection_compute_resources_path) + image_tag('/assets/spinner.gif', :id => 'test_connection_indicator', :class => 'hide').html_safe }) %>
+                 :'data-url' => test_connection_compute_resources_path) + spinner('', :id => 'test_connection_indicator', :class => 'hide').html_safe }) %>
```
does it make sense for spinner to add hide class by default? (or at least an option as pretty common to hide by default)?

ohadlevy commented on app/views/config_reports/welcome.html.erb:
```diff
@@ -1,11 +1,16 @@
-
-  <p><%= _('This page will self destruct once a report comes in.') %></p>
+  <div class="blank-slate-pf-main-action">
+    <%= link_to _('Documentation'), "http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/index.html#3.5.4PuppetReports", :rel => 'external', :class => 'btn btn-primary btn-lg' %>
```
documentation_url

ohadlevy commented on app/views/domains/welcome.html.erb:
```diff
@@ -1,13 +1,18 @@
+    <%= _("The <b>fullname</b> field is used for human readability in reports and other pages that refer to domains,
+    and also available as an external node parameter").html_safe %></br>
   </p>
+  <p><%= _('Learn more about this') %> <%= link_to _('on the documentation'), "http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/#4.4Provisioning"%></p>
```
ditto

ohadlevy commented on app/views/hosts/_progress.html.erb:
```diff
@@ -1,8 +1,8 @@
-  <h3><%= action_name.capitalize %> <%= _('in Progress') %></h3>
-  <p id="tasks_progress"><%= action_name %> in progress.</p>
-  <div class="progress progress-striped active">
+  <h3><span class='pficon pficon-info'></span> <%= action_name.capitalize %> <%= _('in Progress') %></h3>
```
why not use helper?

dLobatog replied:
@alongoldboim ^

alongoldboim replied:
Added helper for it, but there are tons of places we can use icon helper method.
opened an issue: http://projects.theforeman.org/issues/12972

ohadlevy commented on app/views/hosts/_selected_hosts.html.erb:
```diff
@@ -1,42 +1,40 @@
+        <th>
+          <%= _("Environment") %>
+        </th>
+        <%= raw("<th>" + _("Location") + "</th>") if SETTINGS[:locations_enabled]%>
```
why raw?

ohadlevy replied:
nm, i see.

ohadlevy commented on app/views/hosts/welcome.html.erb:
```diff
@@ -1,10 +1,13 @@
-  </p>
-
-  <p><%= _('This page will self destruct once data comes in.') %></p>
+  <p><%= _('Learn more about this') %> <%= link_to _('on the documentation'), "http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/#4.4Provisioning"%></p>
```
documentation helper

ohadlevy commented on app/views/lookup_keys/_value.html.erb:
```diff
@@ -11,11 +11,13 @@
-                    :disabled => !is_param,
-                    :onchange => 'toggleUsePuppetDefaultValue(this, "value")' if is_param %>
-  </td>
+  <% if is_param %>
```
rebase fail?

ohadlevy replied:
?

ohadlevy commented on app/views/lookup_keys/_value.html.erb:
```diff
@@ -11,11 +11,13 @@
+    <td class="ca">
+      <%= f.check_box :use_puppet_default, :'data-property' => 'use_puppet_default',
+                      :disabled => !is_param,
+                      :onchange => 'toggleUsePuppetDefaultValue(this, "value")' if is_param %>
```
oh, you have the same condition twice

ohadlevy commented on app/views/realms/welcome.html.erb:
```diff
@@ -1,8 +1,15 @@
+  <%= _("When a realm is selected for a host, Foreman contacts the relevant realm smart proxy to create an entry for the host and retrieve it's one-time registration password.") %></p>
   <p>
-  <%= _("Foreman supports automatically creating realm entries for new hosts.  When a realm is selected for a host, Foreman contacts the relevant realm smart proxy to create an entry for the host and retrieve it's one-time registration password.").html_safe %>
+    <%= _('Learn more about this') %> <%= link_to _('on the documentation.'), "http://www.theforeman.org/manuals/#{SETTINGS[:version].short}/#4.3.9Realm"%>
```
helper

ohadlevy commented on app/views/templates/_form.html.erb:
```diff
@@ -24,7 +24,7 @@
 
       <% if show_default? %>
-        <%= checkbox_f f, :default, :label=>_('Default'), :help_block => default_template_description %>
+        <%= checkbox_f f, :default, :label=>_('Default'), :help_inline => default_template_description %>
```
why change?

alongoldboim replied:
@dLobatog any reason for that change?

dLobatog replied:
No, I think I wanted to make it one of these little :grey_question: buttons, but it doesn't look like I did. Revert it if you want.

alongoldboim replied:
reverted.

ohadlevy commented on vendor/assets/stylesheets/multi-select.scss:
```diff
@@ -1,3 +1,5 @@
+@import "mixin";
+
```
can you move all of the ms-\* classes from the application.scss here?

dLobatog commented on LICENSE:
```diff
@@ -1105,4 +1106,90 @@ Library.
     incurred by, or claims asserted against, such Contributor by reason
     of your accepting any such warranty or additional liability.
 
+-----------------------------------------------------------
```
Please remove, we don't vendor font-awesome

dLobatog commented:
After digging a bit into why would we get `Sprockets::FileNotFound: couldn't find file 'bootstrap' with type 'application/javascript'`, these are my findings:
- Assets (app.config.assets.paths) are loaded in this order
  - foreman_docker is added automatically
  - bootstrap-sass is added through an initializer
  - `Sprockets::Context` tries to resolve 'bootstrap' path. This is where we get the error.

The bigger issue here is that bootstrap-sass uses [an initializer](https://github.com/twbs/bootstrap-sass/blob/3.2-stable/lib/bootstrap-sass/engine.rb) that is not guaranteed to load before `Sprockets::Context` is initialized.

This means what happens is:
- `Sprockets::Context` is initialized, with an `app.config.assets.paths` that doesn't include bootstrap-sass. See the [`after_initialize` method](https://github.com/rails/sprockets-rails/blob/v2.0.1/lib/sprockets/railtie.rb#L65), line 64 initializes Sprockets.
- bootstrap-sass [runs the initializer task](https://github.com/rails/sprockets-rails/blob/v2.0.1/lib/sprockets/railtie.rb#L65). At this point it's pointless to run it as it won't be added to `Sprockets::Context`.
- [`environment.resolve!`](https://github.com/rails/sprockets/blob/2d30e546872f6f4c5361fe19c2d89b15c5db2dd4/lib/sprockets/context.rb#L82) tries to resolve 'bootstrap' path. However, since environment was initialized without the bootstrap paths, it will not find it and fail.

What I proposed to fix it was to simply, add an initializer [like this](https://gist.github.com/dLobatog/be3d3632e79e25a76320) ourselves to ensure the bootstrap-sass paths are added at a proper time. I believe bootstrap-sass initializer task should run with `:before => :finisher_hook` on, to ensure is always loaded before `Sprockets::Context`.

cc @ehelms @alongoldboim

alongoldboim commented:
@dLobatog Still work in progress toke care of some issues:
- changed icon_text to be similar to icon_text ruby helper, added .use strict
- changed icon_text ruby helper.
- byebug_history removed from gitignore
- icons tag is i.
- spinner text is located to the left
- moved spinner helper to application helper.
- add use of documentation_url helper in the needed views.
- added initialization file to prevent plugins failure.

alongoldboim commented on config/initializers/assets_paths.rb:
```diff
@@ -0,0 +1,5 @@
+  Rails.application.config.assets.paths << Bootstrap::Rails::Engine.root.join('assets', sub).to_s
+  Rails.application.config.assets.paths << Patternfly::Rails::Engine.root.join('assets', sub).to_s
+  Rails.application.config.assets.paths << FontAwesome::Sass::Rails::Engine.root.join('assets', sub).to_s
+end
```
@dLobatog @ohadlevy , This looks to be a decent temp fix for the problem we have.
Looking forward i think we should either do some pull requests for pf,bs,fa sass gems and add a before hook, or change the finisher_hook in our plugins.
Adding bootstrap wasn't enough because fa and pf also use initializer to register their paths.
There are two more initializers im not sure about quiet_assets, and ace-rails, meanwhile i didn't add them, let me know if you think otherwise.

dLobatog commented:
@alongoldboim I refined the initializer a bit. Now what it does it to look for foreman plugins, check if any of them initialize the plugin using `:after => :finisher_hook`, then change it to `:before`. It also prints a deprecation warning. Hopefully this is a bit more user-friendly.

``` ruby
plugins = Rails.application.railties.select { |e| e.railtie_name.match /foreman/ }
plugins.map(&:initializers).flatten.each do |initializer|
  options = initializer.instance_variable_get('@options')

  if options.key(:finisher_hook).present? && options.key(:finisher_hook) == :after
    deprecation_message = "\nInitializing plugins using :after => :finish_hook in\
your engine is deprecated. It delays the load of some assets in Sprockets.\n
Please change your plugin engine initializer to use :before =>\
:finisher_hook or after_initialize."
    Foreman::Deprecation.deprecation_warning('1.13', deprecation_message)
    options[:before] = options.delete :after
    initializer.instance_variable_set('@options', options)
  end
end
```

alongoldboim commented:
@dLobatog :+1:  nice!

dLobatog commented:
@ohadlevy about https://github.com/theforeman/foreman/pull/2943#issuecomment-167999377 - either enlarging the numbers or removing the colors and using plain numbers. I'd vote for the second option based on how they use them here -> [Patternfly/bootstrap docs](https://www.patternfly.org/widgets/#badges).

dLobatog commented:
@alongoldboim A few new issues:
- [ ] Alerts started to look strange, they ought to look like this - https://www.patternfly.org/widgets/#badges
  ![screenshot from 2015-12-31 11-16-59](https://cloud.githubusercontent.com/assets/598891/12063505/3d3d6b1e-afb0-11e5-81af-29b23ec3f7e1.png)
  ![screenshot from 2015-12-31 11-17-10](https://cloud.githubusercontent.com/assets/598891/12063506/3d4102c4-afb0-11e5-92cd-043e62715210.png)
  ![screenshot from 2015-12-31 11-17-18](https://cloud.githubusercontent.com/assets/598891/12063507/3d4157a6-afb0-11e5-9c23-cf8cdf9c83e1.png)
- [ ] Badges text is tiny on /audits 
  ![screenshot from 2015-12-31 11-21-35](https://cloud.githubusercontent.com/assets/598891/12063546/bf4fd3ee-afb0-11e5-88f7-8997aa72d8d3.png)
- [ ] Rubocop violations - http://ci.theforeman.org/job/test_develop_pr_rubocop/5370/checkstyleResult/new/

alongoldboim commented:
@dLobatog 
- Fixed rubocop violations.
- Fixed audit/badges size.
- Submit_or_cancel was placed back in layout.helper
- Trunc with tooltip use was removed.
- Hidden spinner helper was added.
- Opened an issue regarding the use of icon helper (http://projects.theforeman.org/issues/12972)
- Multi-select and dataTables css were extracted to separate files.
- Fixed trends warnings.
- Improved initializer file was added.
- Could you give me some more info about how to generate those bad alerts?
  All the alerts i generated looks good.

ohadlevy commented on LICENSE:
```diff
@@ -54,7 +54,8 @@ Holders for code above are:
   (c) 2010 Giva, Inc.
-  (c) 2012 -2014 Jon Nylander
+  (c) 2012-2014 Jon Nylander
+  (c) 2012-2013 Dave Gandy
```
i dont think you need to add this line (Dave Gandy) as its part of font-awesome package afaiu. @dLobatog ?

ohadlevy commented on LICENSE:
```diff
@@ -1105,4 +1106,3 @@ Library.
     incurred by, or claims asserted against, such Contributor by reason
     of your accepting any such warranty or additional liability.
 
------------------------------------------------------------------------------
```
nitpick: why remove this line?

ohadlevy commented on app/assets/javascripts/application.js:
```diff
@@ -392,24 +392,40 @@ function foreman_url(path) {
-  var options = { type: type, sticky: (type != 'success') };
-  $.jnotify($(item).text(), options);
+  var icon = typeToIcon(type);
+  var options = { classMessage: "alert alert-" + type,
```
note: jnotify does not support multiple classes passed via classMessage, maybe we should simply add the class via jquery instead?

ohadlevy commented on app/assets/javascripts/application.js:
```diff
@@ -558,7 +574,9 @@ function disableButtonToggle(item, explicit) {
+function icon_text(name, inner_text, icon_class) {
   "use strict";
-  return '<span class="glyphicon glyphicon-' + name + '" />';
+  var icon = '<i class="' + icon_class + " " + icon_class + "-" + name + '"/>'
```
@tbrisker i was not sure what is the outcome of the discussion, should we be using `<i>` or `<span>` ?

tbrisker replied:
I'd say use `<span>`, even if font-awesome for some reason chose to use `<i>` in their examples as `<span>` is the correct one according to the HTML specs.

ohadlevy commented on app/assets/stylesheets/application.scss:
```diff
@@ -1,14 +1,14 @@
-
+@import "jquery.ui.theme";
+@import "jquery.ui.core";
+@import "jquery.ui.menu";
```
do we need all of the above just for the autocomplete function? if we do, worth mentioning it, if not, i guess they can be removed?

alongoldboim replied:
we need them, adding comment.

ohadlevy commented on app/assets/stylesheets/application.scss:
```diff
@@ -219,13 +219,19 @@ select {
 
+.form-group > div.control-label{
+  text-align: left;
+  > a{
```
nitpick: whitespace between `a` and `{`

ohadlevy commented on app/assets/stylesheets/application.scss:
```diff
@@ -249,63 +255,11 @@ table {
   white-space: nowrap;
 }
 
+
```
nitpick: extra line

ohadlevy commented on app/assets/stylesheets/application.scss:
```diff
@@ -538,11 +513,43 @@ span.glyphicon.host-status {
+
+.white-header {
+  background-color: white !important;
+  th{
```
nitpick: whitespace, maybe use your browser setting to auto align/correct

ohadlevy commented on app/assets/stylesheets/application.scss:
```diff
@@ -538,11 +513,43 @@ span.glyphicon.host-status {
+  float: left;
+}
+
+@import "/**/*";
```
should this be last? if it serve a specific purpose maybe comment on it so we keep it last?

ohadlevy commented on app/assets/stylesheets/patternfly_and_overrides.scss:
```diff
@@ -239,23 +223,66 @@ textarea {
+  font-size: 100%;
+}
+
+
```
nitpick: extra lines

ohadlevy commented on app/helpers/layout_helper.rb:
```diff
@@ -493,13 +498,23 @@ def new_child_fields_template(form_builder, association, options = { })
     "table table-bordered table-striped table-condensed " + classes
   end
+
+  def submit_or_cancel(f, overwrite = false, args = { })
```
can you add it back to its original location so it doesn't show in the diff? (its hard to tell if it was changed or not)

ohadlevy commented on app/views/architectures/welcome.html.erb:
```diff
@@ -1,14 +1,16 @@
+<% content_for(:title, _("Architectures")) %>
+ <div class="blank-slate-pf">
+  <div class="blank-slate-pf-icon">
+    <i class="fa fa-building-o"></i>
```
use the icon helpers?

ohadlevy commented on app/views/architectures/welcome.html.erb:
```diff
@@ -1,14 +1,16 @@
   </p>
+  <p><%= _('Learn more about this') %> <%= link_to _('on the documentation'), documentation_url("4.4.5Architectures")%></p>
+  <div class="blank-slate-pf-main-action">
+      <%= link_to(_("New architecture"), new_architecture_path, :class => "btn btn-primary btn-lg") %>
```
afair the classes are added via javascript on all new btns?

tbrisker replied:
This is outside the scope of this PR, but just wanted to comment that the face we use js to add the classes to all links is not good practice - js should deal with interactions, not with styling that should have been done by the server.

ohadlevy replied:
@tbrisker, agree it should simply be a helper.

ohadlevy commented on app/views/auth_source_ldaps/welcome.html.erb:
```diff
@@ -1,8 +1,15 @@
-  <p> <%= _("The authentication process currently requires an LDAP provider, such as <em>OpenLDAP</em> or <em>Microsoft's Active Directory</em>.").html_safe %></p>
+<div class="blank-slate-pf">
+  <div class="blank-slate-pf-icon">
+    <i class="fa fa-users"></i>
```
icon helper

ohadlevy commented on app/views/auth_source_ldaps/welcome.html.erb:
```diff
@@ -1,8 +1,15 @@
+  </p>
+  <p><%= _('Learn more about this') %> <%= link_to _('on the documentation'), documentation_url("4.1.1LDAPAuthentication")%></p>
+  <div class="blank-slate-pf-main-action">
+    <%= link_to(_("New authentication source"), new_auth_source_ldap_path, :class => 'btn btn-primary btn-lg') %>
```
ditto

ohadlevy commented on app/views/bookmarks/welcome.html.erb:
```diff
@@ -1,8 +1,12 @@
+<% content_for(:title, _("Bookmarks")) %>
+<div class="blank-slate-pf">
+  <div class="blank-slate-pf-icon">
+    <i class="fa fa-bookmark"></i>
```
ditto

ohadlevy commented on app/views/bookmarks/welcome.html.erb:
```diff
@@ -1,8 +1,12 @@
+  <p><%= _("You don't seem to have any bookmarks.") %></p>
+  <p><%= _("Bookmarks are saved search queries. To save them, click on 'Bookmark this search' on the dropdown next to every search button.") %></p>
+  <div class="blank-slate-pf-main-action">
+      <%= link_to _('Documentation'), documentation_url("4.1.5Searching"), :rel => "external", :class => "btn btn-primary btn-lg" %>
```
here too

ohadlevy commented on app/views/common_parameters/index.html.erb:
```diff
@@ -3,22 +3,23 @@
 
-<table class="table table-bordered table-striped table-two-pane table-fixed">
+
+<table class="table table-bordered table-striped table-two-pane table-hover table-fixed">
```
why not use the table css helper?

ohadlevy replied:
missed this one?

ohadlevy commented on app/views/config_reports/welcome.html.erb:
```diff
@@ -1,11 +1,16 @@
-<div id="welcome">
+<div class="blank-slate-pf">
+  <div class="blank-slate-pf-icon">
+    <i class="fa fa-book"></i>
```
helper

ohadlevy commented on app/views/config_reports/welcome.html.erb:
```diff
@@ -1,11 +1,16 @@
-
-  <p><%= _('This page will self destruct once a report comes in.') %></p>
+  <div class="blank-slate-pf-main-action">
+    <%= link_to _('Documentation'), documentation_url("3.5.4PuppetReports"), :rel => 'external', :class => 'btn btn-primary btn-lg' %>
```
btn

ohadlevy commented on app/views/dashboard/_reports_widget.html.erb:
```diff
@@ -5,7 +5,7 @@
   <p class="ca"><%= _("No interesting reports received in the last week") %></p>
 <% else %>
-  <table class="table table-striped ellipsis">
+  <table class="table table-striped table-bordered ellipsis">
```
css table helper?

ohadlevy commented on app/views/domains/welcome.html.erb:
```diff
@@ -1,13 +1,18 @@
-<div id="welcome">
+<div class="blank-slate-pf">
+  <div class="blank-slate-pf-icon">
+    <i class="fa fa-globe"></i>
```
helper

ohadlevy commented on app/views/domains/welcome.html.erb:
```diff
@@ -1,13 +1,18 @@
   </p>
+  <p><%= _('Learn more about this') %> <%= link_to _('on the documentation'), documentation_url("4.4Provisioning")%></p>
+  <div class="blank-slate-pf-main-action">
+      <%= link_to _("New Domain"), new_domain_path, :class => "btn btn-primary btn-lg" %>
```
btn

ohadlevy commented on app/views/environments/welcome.html.erb:
```diff
@@ -1,16 +1,14 @@
+<% content_for(:title, _("Puppet environments")) %>
+<div class="blank-slate-pf">
+  <div class="blank-slate-pf-icon">
+    <i class="fa fa-th"></i>
```
helper

ohadlevy commented on app/views/environments/welcome.html.erb:
```diff
@@ -1,16 +1,14 @@
+  <%= _('Environments may be manually created and only require the name of the environment to be declared.') %></br>
+  <p><%= _('Learn more about this') %> <%= link_to(_('on the documentation'), documentation_url("4.2.1Environments")) %></p>
+  <div class="blank-slate-pf-main-action">
+      <%= link_to _('New Puppet environment'), new_environment_path, :class => "btn btn-primary btn-lg" %>
```
btn

ohadlevy commented on app/views/fact_values/index.html.erb:
```diff
@@ -1,7 +1,7 @@
 <% title_actions documentation_button('3.5.5FactsandtheENC') %>
 
-<table class="table table-bordered table-striped">
+<table class="table table-bordered table-striped ellipsis">
```
css table helper

ohadlevy commented on app/views/fact_values/welcome.html.erb:
```diff
@@ -1,9 +1,12 @@
-  <p><%= _('This page will self destruct once inventory data comes in.') %></p>
+<% content_for(:title, _("Fact values")) %>
+<div class="blank-slate-pf">
+  <div class="blank-slate-pf-icon">
```
helper

ohadlevy commented on app/views/fact_values/welcome.html.erb:
```diff
@@ -1,9 +1,12 @@
+  <%= _("You don't seem to have any facts yet. If you wish to configure fact pushing, please follow the documentation.") %>
+  <div class="blank-slate-pf-main-action">
+    <%= link_to _('Documentation'), documentation_url("3.5.5FactsandtheENC"), :rel => "external",
+      :class => "btn btn-primary btn-lg" %>
```
btn

ohadlevy commented on app/views/locations/welcome.html.erb:
```diff
@@ -1,9 +1,15 @@
-      <p><strong><%= _("Notice") %></strong> <%= _("You must create a location before continuing.") %></p>
+<div class="blank-slate-pf">
+  <div class="blank-slate-pf-icon">
+    <i class="fa fa-globe"></i>
```
helper

ohadlevy commented on app/views/locations/welcome.html.erb:
```diff
@@ -1,9 +1,15 @@
+    <%= (_('Locations also work nicely when used in tandem with %{organizations}.') % {:organizations => link_to(_("organizations"), :controller => "organizations", :action => "index")}).html_safe %>
+  </p>
+  <div class="blank-slate-pf-main-action">
+      <%= link_to(_("New location"), new_location_path, :class => "btn btn-primary btn-lg") %>
```
btn

ohadlevy commented on app/views/organizations/welcome.html.erb:
```diff
@@ -1,13 +1,15 @@
-    </p>
+<div class="blank-slate-pf">
+  <div class="blank-slate-pf-icon">
+    <i class="fa fa-users"></i>
```
helper

ohadlevy commented on app/views/organizations/welcome.html.erb:
```diff
@@ -1,13 +1,15 @@
+    <%= (_('Organizations also work nicely in tandem with %{locations}.') % {:locations => link_to(_("locations"), :controller => "locations", :action => "index")}).html_safe %>
+  </p>
+  <div class="blank-slate-pf-main-action">
+    <%= link_to(_("New organization"), new_organization_path, :class => 'btn btn-primary btn-lg') %>
```
btn

ohadlevy commented on app/views/realms/welcome.html.erb:
```diff
@@ -1,8 +1,15 @@
+<% content_for(:title, _("Realms")) %>
+<div class="blank-slate-pf">
+  <div class="blank-slate-pf-icon">
+    <i class="fa fa-key"></i>
```
helper

ohadlevy commented on app/views/realms/welcome.html.erb:
```diff
@@ -1,8 +1,15 @@
+    <%= _('Learn more about this') %> <%= link_to _('on the documentation.'), documentation_url("4.3.9Realm")%>
   </p>
+  <div class="blank-slate-pf-main-action">
+    <%= display_link_if_authorized _('New Realm'), hash_for_new_realm_path, :class => "btn btn-primary btn-lg" %>
```
btn

ohadlevy commented on app/views/usergroups/_form.html.erb:
```diff
@@ -19,12 +19,12 @@
     <% if AuthSource.non_internal.present? %>
       <div class="tab-pane" id="external">
-        <table class="table table-bordered table-striped table-two-pane">
+        <table class="table table-bordered table-striped table-two-pane table-hover">
```
css table helper?

alongoldboim commented:
@ohadlevy @dLobatog 
- removed vendored jsnotify change, used jquery to replace.
- fixed nitpicks
- changed welcome pages to icons to use icon_text helper, changed helpers to use span instead of i tag
- add issues that need to be addressed once this pr is merged:
  http://projects.theforeman.org/issues/12972
  http://projects.theforeman.org/issues/12945
  http://projects.theforeman.org/issues/12979

ohadlevy commented on config/initializers/assets_paths.rb:
```diff
@@ -0,0 +1,14 @@
+plugins = Rails.application.railties.select { |e| e.railtie_name.match /foreman/ }
```
imho worth a comment about what this is all about

alongoldboim replied:
added comment.

ohadlevy commented on bundler.d/assets.rb:
```diff
@@ -1,7 +1,10 @@
+gem 'autoprefixer-rails', '~> 5.2'
+gem 'sprockets', '~> 3'
+gem 'patternfly-sass', '~> 2.7'
+gem 'sprockets-rails', '~> 2.3.3'
```
note this line needs to be rebased with develop

dLobatog commented on config/initializers/assets_paths.rb:
```diff
@@ -0,0 +1,20 @@
+=begin
```
@alongoldboim Use plain old # comments, rubocop is failing [because of this block](https://github.com/bbatsov/ruby-style-guide#no-block-comments)

dLobatog commented:
@alongoldboim The test failure seems like it's not caused by this. I can fix the rubocop violation & use the table css helper class that @ohadlevy recommended on merge.

alongoldboim commented:
@dLobatog talked with @ohadlevy , agreed on applying the use of table_css helper app wide on a diff pr.

dLobatog commented:
Merged as aa50ab15ff091f132be81861f03c6a9919f2410d, thanks @alongoldboim for picking this one up, and @ohadlevy @ehelms @shlomizadok @tbrisker for doing the reviews.
